upgrade to Tokio v1.x ecosystem

tdyas opened the pull request:
### Problem

Stay current on Tokio / Tonic / Prost / Hyper ecosystem.

### Solution

Upgrade to the Tokio v1.x ecosystem. A minimum of Tokio v1.4 is required in order to have `Handle::block_on` available for use by the `task_executor` crate.

### Result

Existing tests pass.

tdyas commented:
This compiles but has some tests failing in `process_execution`. There are some questions that need answering before proceeding to finalize this. Will tag appropriate people.

tdyas commented:
Prerequisite is a release of the `nails` crate with Tokio 1.0 ecosystem support from https://github.com/stuhood/nails/pull/2.

tdyas commented on src/rust/engine/nailgun/src/tests.rs:
```diff
@@ -47,8 +46,12 @@ async fn shutdown_awaits_ongoing() {
     move |_| {
-      connection_accepted.notify();
-      Handle::current().block_on(should_complete_connection.notified());
+      connection_accepted.notify_one();
+      loop {
+        if let Some(_) = should_complete_connection.notified().now_or_never() {
+          break;
+        }
+      }
```
`block_on` is only available on `Runtime` now and not `Handle`. For now, I have written a busy loop to poll the `Future`, but how we handle blocking work in Tokio 1.0 needs some thought. This busy loop should be changed before the final version of this PR.

cc @stuhood

stuhood replied:
Yea, this blocks the upgrade. See https://github.com/tokio-rs/tokio/issues/2965... have pinged it.

stuhood replied:
`block_on` is now un...blocked! https://github.com/tokio-rs/tokio/issues/3096#issuecomment-803371380

tdyas replied:
Done.

tdyas commented on src/rust/engine/process_execution/Cargo.toml:
```diff
@@ -10,25 +10,25 @@ async-trait = "0.1"
 libc = "0.2.39"
 log = "0.4"
-nails = "0.11"
+nails = { git = "https://github.com/tdyas/nails", rev = "889abe4e63a27899d2f6ce50adc8489b289a19b0" }
```
Update whenever `nails` is released with Tokio 1.0 support.

stuhood replied:
Now released as `0.12.0`.

tdyas commented on src/rust/engine/process_execution/src/local.rs:
```diff
@@ -275,6 +278,27 @@ impl super::CommandRunner for CommandRunner {
   }
 }
 
+use ouroboros::self_referencing;
+
+#[self_referencing]
+struct OwnedChild {
+  child: Box<Child>,
+  #[borrows(mut child)]
+  #[not_covariant]
+  exit_stream: BoxStream<'this, Result<ChildOutput, std::io::Error>>,
+}
+
+impl Stream for OwnedChild {
+  type Item = Result<ChildOutput, std::io::Error>;
+
+  fn poll_next(
+    mut self: Pin<&mut Self>,
+    cx: &mut std::task::Context<'_>,
+  ) -> Poll<Option<Self::Item>> {
+    self.with_exit_stream_mut(|es| Pin::new(es).poll_next(cx))
+  }
+}
+
```
`tokio::process::Child` no longer implements `Future` in Tokio 1.0. Instead, `Child` exposes a `wait` instance method. The problem is that `wait` takes `&mut self` and so the original version of `exit_stream` thus mutably borrows `self` and so cannot be returned from `run_in_workdir` since the borrow checker rejects the code since `Child` does not survive the call to `run_in_workdir`.

The solution is to have `Child` survive that call. This code uses the `ouroboros` crate which has a proc macro that allows for a limited form of self-referential structs. The `Child` struct is owned by this struct and this struct is what comprises part of the stream returned from `run_in_workdir`. The `exit_stream` part of the return value from `run_in_workdir` is also embedded in this struct using the `ouroboros` crate.

Finally, the solution implements `Stream` on `OwnedChild` so that it exposes the same stream as `exit_stream` but as a first-class owner of the embedded `Child`.

cc @stuhood

stuhood replied:
So, afaik, the suggested way to do this is to call `child.stdout.take()` (which is identical to how you do this with the `stdlib` `Child` type). See the `stdout`/`stderr` fields in the stdlib docs https://doc.rust-lang.org/std/process/struct.Child.html for more info. You can then call `wait` safely.

tdyas replied:
The issue is not with `child.stdout` nor with `child.stderr`. The `stdout_stream` and `stderr_stream` streams take ownership of those respective values and are not triggering the borrow checker. See the construction of `stdout_stream` and `stderr_stream` where the code already calls `child.stdout.take().unwrap()` and `child.stderr.take().unwrap()` respectively.

The issue is with `exit_stream`. In tokio v0.2.x, `Child` implements `Future` and can be used directly to await the exit status of the child process. The borrow checker was not triggered in `run_in_workdir` previously because `exit_stream` took ownership of the underlying `Future` a/k/a the `Child` struct. Thus returning `exit_stream` from `run_in_workdir` implicitly passed ownership of `Child` out of `run_in_workdir` because `exit_stream` owned it.

In tokio v1.x, that is not the case, `Child` no longer implements `Future` and one must call `.wait` to obtain a Future that can be used to await the child's exit status. `.wait` takes `&mut self` and thus borrows `Child` mutably within the call to `run_in_workdir`. Unlike in tokio v0.2.x, the borrow checker triggers in this case because `exit_stream` now only has the `&mut self` reference to `Child` (via `.wait`) and not ownership of `Child` like it did previously. Thus the borrow checker triggers because the `Child` is dropped when `run_in_workdir` returns.

The solution is to let `Child` escape `run_in_workdir`. However, the code should also keep the Future returned from `.wait` alive (and its `&mut self` reference) so that we only construct `exit_stream` once. This is accomplished by using the `ouroboros` crate to safely maintain the self-reference between `Child` and `exit_stream`.

And to make things easier, the `OwnedChild` struct implements Stream so that the details are hidden of how `exit_stream` and the `Child` struct are stored.

stuhood replied:
Hm. Ok. I expect that moving forward we should move away from the "unified stream of stdout and exit code" interface here, as TODO'd in: https://github.com/pantsbuild/pants/pull/11370/files#r547623793. `nails` did to ease lifetime issues, and it seemed to work fine. Happy to land this as is if we preserve the TODO though.

tdyas replied:
Fine by me. I'll preserve the TODO.

tdyas commented on src/rust/engine/process_execution/src/local.rs:
```diff
@@ -361,27 +385,39 @@ impl CapturedWorkdir for CommandRunner {
+      child: Box::new(child),
+      exit_stream_builder: |child: &mut Child| {
+        child
+          .wait()
```
Here is the mutable borrow of `child`.

tdyas replied:
Huh, using a closure like that works. I had not been wrapping it in an async block, which causes the borrow checker  to fail with:

```
   --> process_execution/src/local.rs:391:5
    |
376 |         child
    |         ----- `child` is borrowed here
...
391 | /     Ok(
392 | |       result_stream
393 | |         .map_err(|e| format!("Failed to consume process outputs: {:?}", e))
394 | |         .boxed(),
395 | |     )
    | |_____^ returns a value referencing data owned by the current function

error[E0515]: cannot return value referencing local variable `child`
```

The joys of Rust borrowing rules ... <sigh>

Updated to not use the `ouroboros` crate.

tdyas commented on src/rust/engine/task_executor/src/lib.rs:
```diff
@@ -61,11 +61,11 @@ impl Executor {
-      .core_threads(core_threads)
-      .max_threads(num_threads)
-      .threaded_scheduler()
+  // TODO(tokio-1.0): Figure out thread count configuration in Pants.
+  pub fn new_owned(_core_threads: usize, num_threads: usize) -> Result<Executor, String> {
+    let runtime = Builder::new_multi_thread()
+      .worker_threads(num_threads)
+      .max_blocking_threads(num_threads)
```
The thread configuration for the multi-thread runtime changed in v1.0. Tokio no longer dynamically scales the number of threads as it does in prior versions. Instead, it will just spin up `worker_threads` threads and ensure that only `max_blocking_threads` are used for blocking work. (At least that is my naive take on the matter.)

The questions are:

1. What API should we be exposing for PyExecutor and related types?

2. The default number of threads now is the number of available cores. Should we just trust the Tokio defaults now and not perform manual configuration in the general case?

stuhood replied:
According to the discussion on https://github.com/tokio-rs/tokio/issues/2802 this is now `worker_threads` + `max_blocking_threads`? Commented to confirm though: https://github.com/tokio-rs/tokio/issues/2802#issuecomment-763889643

tdyas replied:
How should we express the concept of letting Tokio choose the `worker_threads` value while still retaining the ability to configure the number of blocking threads? Just do `max-core` and only use that value?

Darksonn replied:
> Tokio no longer dynamically scales the number of threads as it does in prior versions.

To be clear, the only thing that has changed here is how you set the upper limit on blocking threads. Just like in the previous versions, the number of core threads is fixed, and the number of blocking threads is dynamicly scaled until it reaches the limit.

tdyas commented:
cc @stuhood @Eric-Arellano @illicitonion 

Low priority. The PR has some questions to be answered in the coming week, which will require some thought on our part, so not marking for review yet.

tdyas commented on src/rust/engine/task_executor/src/lib.rs:
```diff
@@ -119,7 +120,9 @@ impl Executor {
-      .handle
+      .runtime
+      .as_ref()
+      .unwrap()
```
This is returning `None` and causing some tests to fail.

stuhood replied:
Mm. Yea, this is core to the `block_on` discussion, I think. The `handle` is a weak-reference to the `Runtime`, which doesn't prevent shutdown, and so it can fail.

stuhood commented:
`nails` `0.12.0` is released, using `tokio` `1.x`: https://crates.io/crates/nails/0.12.0

stuhood commented on src/rust/engine/async_latch/src/lib.rs:
```diff
@@ -72,7 +72,7 @@ impl AsyncLatch {
     // To see whether the latch is triggered, we clone the receiver, and then wait for our clone to
     // return None, indicating that the Sender has been dropped.
```
Mind updating this?

tdyas replied:
I missing context. In what way?

stuhood replied:
"then wait for our clone to return None" is no longer accurate.

tdyas replied:
Fixed.

stuhood reviewed (commented):
Thanks a lot!

tdyas commented:
@stuhood: Regarding the use of `block_on`, do we have a definitive list of the contexts where Pants uses `block_on`? And is it the case that those call sites don't have access to a `Runtime` or even a Handle to use spawn_blocking or `tokio::task::block_in_place` (assuming it is relevant)?

 I'm trying to see if there is a way to work around the lack of `block_on` on Handle.

tdyas commented:
Rebased and update PR for latest `master`. Apparently https://github.com/tokio-rs/tokio/pull/3569 will add `block_on` back to `Handle`, but has not landed in Tokio yet.

tdyas commented:
Rebased again and updated Tokio to be a minimum of v1.4 which includes the change to add `block_on` back to Handle. (This is also all squashed down to a single commit. I recommend reviewing anew.)

stuhood reviewed (approved):
Thanks! One `block_on` to restore, but pre-shipping for once that is fixed.

tdyas commented on src/rust/engine/async_latch/Cargo.toml:
```diff
@@ -7,7 +7,7 @@ publish = false
 
 [dev-dependencies]
-tokio = { version = "0.2.23", features = ["macros"] }
+tokio = { version = "^1.4", features = ["rt", "macros", "time"] }
```
Fixed.

stuhood reviewed (approved):
Thanks again!
